sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -143,145 +115,226 @@ private[sql] object OrcFilters extends OrcFiltersBase {
     case _ => value
   }
 
+  import org.apache.spark.sql.sources._
+  import OrcFilters._
+
   /**
-   * Build a SearchArgument and return the builder so far.
+   * Builds a SearchArgument for a Filter by first trimming the non-convertible nodes, and then
+   * only building the remaining convertible nodes.
+   *
+   * Doing the conversion in this way avoids the computational complexity problems introduced by
+   * checking whether a node is convertible while building it. The approach implemented here has
+   * complexity that's linear in the size of the Filter tree - O(number of Filter nodes) - we run
+   * a single pass over the tree to trim it, and then another pass on the trimmed tree to convert
+   * the remaining nodes.
+   *
+   * The alternative approach of checking-while-building can (and did) result
+   * in exponential complexity in the height of the tree, causing perf problems with Filters with
+   * as few as ~35 nodes if they were skewed.
    */
-  private def buildSearchArgument(
-      dataTypeMap: Map[String, DataType],
+  private[sql] def buildSearchArgument(
       expression: Filter,
       builder: Builder): Option[Builder] = {
-    createBuilder(dataTypeMap, expression, builder, canPartialPushDownConjuncts = true)
+    trimUnconvertibleFilters(expression).map { filter =>
+      updateBuilder(filter, builder)
+      builder
+    }
+  }
+
+  private[sql] def trimUnconvertibleFilters(expression: Filter): Option[Filter] = {
```

Comment:
nit: since `updateBuilder` has doc, it's better to add doc for this one as well.